File: csv_parser.go
// csv_parser.go — Go версия

package main

import (
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

func detectDelimiter(filePath string) rune {
	file, _ := os.Open(filePath)
	defer file.Close()
	reader := csv.NewReader(file)
	firstLine, _ := reader.Read()
	if len(firstLine) == 0 {
		return ','
	}
	// Пробуем определить по количеству полей
	for _, delim := range []rune{',', ';', '\t', '|'} {
		r := csv.NewReader(strings.NewReader(strings.Join(firstLine, string(delim))))
		fields, _ := r.Read()
		if len(fields) > 1 {
			return delim
		}
	}
	return ','
}

func parseValue(value string) interface{} {
	value = strings.TrimSpace(value)
	if value == "" {
		return nil
	}
	switch strings.ToLower(value) {
	case "true":
		return true
	case "false":
		return false
	case "null", "none":
		return nil
	}
	if i, err := strconv.Atoi(value); err == nil {
		return i
	}
	if f, err := strconv.ParseFloat(value, 64); err == nil {
		return f
	}
	return value
}

func parseNestedKeys(headers []string) map[string]interface{} {
	result := make(map[string]interface{})
	for _, key := range headers {
		parts := strings.Split(key, ".")
		current := result
		for _, part := range parts[:len(parts)-1] {
			if _, ok := current[part]; !ok {
				current[part] = make(map[string]interface{})
			}
			current = current[part].(map[string]interface{})
		}
		current[parts[len(parts)-1]] = nil
	}
	return result
}

func buildNestedRecord(row []string, headers []string) map[string]interface{} {
	record := make(map[string]interface{})
	for i, header := range headers {
		parts := strings.Split(header, ".")
		current := record
		for _, part := range parts[:len(parts)-1] {
			if _, ok := current[part]; !ok {
				current[part] = make(map[string]interface{})
			}
			current = current[part].(map[string]interface{})
		}
		val := ""
		if i < len(row) {
			val = row[i]
		}
		current[parts[len(parts)-1]] = parseValue(val)
	}
	return record
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: go run csv_parser.go <input.csv> [output.json]")
		os.Exit(1)
	}

	inputFile := os.Args[1]
	outputFile := "output.json"
	if len(os.Args) > 2 {
		outputFile = os.Args[2]
	}

	fmt.Println("📊 CSV → JSON Парсер (Go)")

	file, err := os.Open(inputFile)
	if err != nil {
		fmt.Printf("❌ Ошибка: %v\n", err)
		os.Exit(1)
	}
	defer file.Close()

	delim := detectDelimiter(inputFile)
	fmt.Printf("🔍 Определён разделитель: %q\n", delim)

	reader := csv.NewReader(file)
	reader.Comma = delim
	reader.FieldsPerRecord = -1

	headers, err := reader.Read()
	if err != nil {
		fmt.Printf("❌ Ошибка чтения заголовков: %v\n", err)
		os.Exit(1)
	}
	fmt.Printf("📋 Заголовки: %v\n", headers)

	var data []map[string]interface{}
	for {
		row, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			fmt.Printf("⚠️ Ошибка строки: %v\n", err)
			continue
		}
		record := make(map[string]interface{})
		for i, header := range headers {
			val := ""
			if i < len(row) {
				val = row[i]
			}
			record[header] = parseValue(val)
		}
		data = append(data, record)
	}

	jsonData, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		fmt.Printf("❌ Ошибка JSON: %v\n", err)
		os.Exit(1)
	}

	err = os.WriteFile(outputFile, jsonData, 0644)
	if err != nil {
		fmt.Printf("❌ Ошибка сохранения: %v\n", err)
		os.Exit(1)
	}

	fmt.Printf("✅ Экспортировано %d записей в %s\n", len(data), outputFile)
}

File: csv_parser.py


### 1. `csv_parser.py` (Python)

```python
# csv_parser.py — Python версия

import csv
import json
import sys
import os
import re
from datetime import datetime

def detect_delimiter(file_path):
    """Автоопределение разделителя"""
    with open(file_path, 'r', encoding='utf-8-sig') as f:
        first_line = f.readline()
        if '\t' in first_line:
            return '\t'
        elif ';' in first_line:
            return ';'
        else:
            return ','

def parse_value(value):
    """Умное преобразование типов"""
    value = value.strip()
    if value == '':
        return None
    if value.lower() == 'true':
        return True
    if value.lower() == 'false':
        return False
    if value.lower() == 'null' or value.lower() == 'none':
        return None
    # Число
    try:
        if '.' in value:
            return float(value)
        else:
            return int(value)
    except ValueError:
        pass
    # Дата (попытка)
    try:
        for fmt in ['%Y-%m-%d', '%d-%m-%Y', '%Y/%m/%d']:
            try:
                return datetime.strptime(value, fmt).isoformat()
            except ValueError:
                continue
    except:
        pass
    return value

def parse_nested_keys(header):
    """Обработка точечной нотации: user.name -> {'user': {'name': ...}}"""
    result = {}
    for key in header:
        parts = key.split('.')
        current = result
        for part in parts[:-1]:
            if part not in current:
                current[part] = {}
            current = current[part]
        current[parts[-1]] = None
    return result

def build_nested_dict(row, keys):
    """Сборка вложенного словаря из плоского списка значений"""
    result = {}
    for i, key in enumerate(keys):
        parts = key.split('.')
        current = result
        for part in parts[:-1]:
            if part not in current:
                current[part] = {}
            current = current[part]
        current[parts[-1]] = parse_value(row[i]) if i < len(row) else None
    return result

def parse_csv(input_file, output_file=None, delimiter=None, nested=False):
    if not os.path.exists(input_file):
        print(f"❌ Файл {input_file} не найден.")
        return False

    if delimiter is None:
        delimiter = detect_delimiter(input_file)
        print(f"🔍 Определён разделитель: {repr(delimiter)}")

    data = []
    with open(input_file, 'r', encoding='utf-8-sig') as f:
        reader = csv.reader(f, delimiter=delimiter)
        rows = list(reader)
        if not rows:
            print("❌ CSV пуст.")
            return False
        headers = [h.strip() for h in rows[0]]
        print(f"📋 Заголовки: {headers}")

        for row in rows[1:]:
            if nested:
                # Вложенная структура
                # Создаём пустую структуру
                record = {}
                for i, header in enumerate(headers):
                    parts = header.split('.')
                    current = record
                    for part in parts[:-1]:
                        if part not in current:
                            current[part] = {}
                        current = current[part]
                    current[parts[-1]] = parse_value(row[i]) if i < len(row) else None
                data.append(record)
            else:
                # Плоская структура
                record = {}
                for i, header in enumerate(headers):
                    record[header] = parse_value(row[i]) if i < len(row) else None
                data.append(record)

    if output_file is None:
        output_file = os.path.splitext(input_file)[0] + '.json'

    with open(output_file, 'w', encoding='utf-8') as f:
        json.dump(data, f, indent=2, ensure_ascii=False)

    print(f"✅ Экспортировано {len(data)} записей в {output_file}")
    return True

def main():
    if len(sys.argv) < 2:
        print("Usage: python csv_parser.py <input.csv> [output.json] [--nested]")
        print("  --nested  : использовать вложенную структуру из точечной нотации")
        sys.exit(1)

    input_file = sys.argv[1]
    output_file = None
    nested = False

    for arg in sys.argv[2:]:
        if arg == '--nested':
            nested = True
        elif not arg.startswith('--'):
            output_file = arg

    print("📊 CSV → JSON Парсер (Python)")
    parse_csv(input_file, output_file, nested=nested)

if __name__ == "__main__":
    main()
